lib/Sema/CSDiag.cpp

```diff
@@ -3659,10 +3684,33 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Re-type-check the argument with the expected type of the candidate set.
     // This should produce a specific and tailored diagnostic saying that the
     // type mismatches with expectations.
-    if (!typeCheckChildIndependently(badArgExpr,
-                                     calleeInfo.failedArgument.parameterType,
+    Type paramType = calleeInfo.failedArgument.parameterType;
+    if (!typeCheckChildIndependently(badArgExpr, paramType,
                                      CTP_CallArgument, options))
       return true;
+      
+    // If that fails, it could be that the argument doesn't conform to an archetype.
+    Type argType = badArgExpr->getType();
+
+    if (paramType->is<ArchetypeType>() && !argType->hasTypeVariable() &&
```

Comment:
Should (or can) this be moved into diagnoseSimpleErrors?  Doing so would get coverage of this in the visitUnresolvedMemberExpr path and the path used by SubscriptExprs.